[SPARK-7826][CORE] Suppress extra calling getCacheLocs.

ueshin opened the pull request:
There are too many extra call method `getCacheLocs` for `DAGScheduler`, which includes Akka communication.
To improve `DAGScheduler` performance, suppress extra calling the method.

In my application with over 1200 stages, the execution time became 3.8 min from 8.5 min with my patch.

ueshin committed:
Suppress extra calling getCacheLocs.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33339/consoleFull) for   PR 6352 at commit [`9a80fad`](https://github.com/apache/spark/commit/9a80fad6b2f5d73e206aacf72cd5e075fa8bcba0).

srowen commented:
Can you explain why it's valid to proceed without the call when there is 1 dependency?
Also it looks like you're adding calls to `getCacheLocs` actually. I don't see an explanation and the description isn't consistent with the change.

SparkQA commented:
[Test build #33339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33339/consoleFull) for   PR 6352 at commit [`9a80fad`](https://github.com/apache/spark/commit/9a80fad6b2f5d73e206aacf72cd5e075fa8bcba0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33339/
Test PASSed.

ueshin commented:
@srowen, Thank you for your checking.

1. To check the parent stages are missing or not, we only have to check the location of RDD that has `ShuffleDependency` because narrow-depending RDDs would be the same location, so I moved the calling `getCacheLocs` into the `case shufDep`.
2. I was afraid that the dependency graph would become bigger (if there are a lot of `union` or `zipPartitions`, etc.) so I added checking the location at the point the RDD has more than 1 dependencies.
  * we might not need [line 389](https://github.com/apache/spark/pull/6352/files#diff-6a9ff7fb74fd490a50462d45db2d5e11R389) if we don't have to consider the case.

srowen commented:
Is part 2 really just to be safe? It seems essential. Are you saying that only shuffle dependencies have more than 1 dependency?  Also this adds a new call to all dependencies. Doesn't this mostly defeat the purpose? I am not an expert on this code but I am not sure the logic is clear here

ueshin commented:
Calling will occur not for all RDDs in the stage but only when:

1. RDD has `ShuffleDependency`.
2. RDD has more than 1 dependencies regardless of the dependency type.

And I should have mentioned that calling for the same RDD is not a problem because the location is already cached.

JoshRosen commented:
(Warning: drive-by comment; I'll look at this patch in more detail later)

One high-level comment:

For any patch which modifies scheduler internals, we should err on the side of extremely liberal commenting of code, even if this means paragraph-long comments.  If it's tricky enough to merit a question in a GitHub code review, then it deserves a comment.  For instance, the `rdd.dependencies.size < 2` check could benefit from a nearby comment that explains why this is safe.

ueshin commented:
Oops, I found that I misunderstood what the method `getCacheLocs` is doing here.
I'll change the way to suppress Akka communications in the next push, so please check this PR after that.

@JoshRosen, Thank you for your comment.
I'll add comments in the next push.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -386,13 +386,15 @@ class DAGScheduler(
     def visit(rdd: RDD[_]) {
       if (!visited(rdd)) {
         visited += rdd
-        if (getCacheLocs(rdd).contains(Nil)) {
```
As a general aside, I find `getCacheLocs(rdd).contains(Nil)` to be hard to understand to begin with.  I think that this condition is meant to be read as "if at least one partition of this RDD is not cached anywhere...".  Maybe this code would be easier to review / parse if we extracted this condition into a variable, perhaps a lazy val if we want to short-circuit, named `rddHasUncachedPartitions`, or `!rddIsCached`.

JoshRosen commented:
Oh, one other thought: maybe a good exercise would be to attempt to write the Scaladoc comment for `getMissingParentStages` which describes, in prose, the basic high-level algorithm for finding missing parent stages.  I can help with this tomorrow.  Even if you don't end up modifying `getMissingParentStages`, I'd love to submit a new PR that just comments / explains the existing code in order to make this easier to understand in the future.

To help me build some intuition for understanding your optimization here:

It looks like this only save us from performing `getCacheLocs` lookups in cases where we're traversing backwards through a long chain of narrow dependencies.  I don't think that this is necessarily safe.  Imagine that we have a lineage graph which looks something like this:

```
┌───┐ shuffle ┌───┐    ┌───┐          
│ A │◀ ─ ─ ─ ─│ B │◀───│ C │◀─┐       
└───┘         └───┘    └───┘  │  ┌───┐
                              ├──│ E │
                       ┌───┐  │  └───┘
                       │ D │◀─┘       
                       └───┘              
```

Here, `E` has one-to-one dependencies on `C` and `D`.  `C` is derived from `A` by performing a shuffle and then a map.  If we're trying to determine which ancestor stages need to be computed in order to compute `E`, we need to figure out whether the shuffle `A -> B` should be performed.  If the RDD `C`, which has only one ancestor via a narrow dependency, is cached, then we won't need to compute `A`, even if it has some unavailable output partitions.  The same goes for `B`: if `B` is 100% cached, then we can avoid the shuffle on `A`.  Based on this, I don't think that we can make a local decision to skip the caching check based on the structure of the RDD graph.  However, we _might_ be able to skip / optimize this check based on RDDs' storage levels: in long chains of narrow dependencies, most RDDs probably _aren't_ cached, so adding a simple `if StorageLevel = None return Seq.fill(numPartitions)(Nil)` check to `getCacheLocs` might be safe / sufficient.

Someone more familiar with StorageLevel / caching semantics should double-check this reasoning to make sure that I'm not overlooking any corner-cases when RDDs' storage levels change due to unpersist / cache / persist calls.

JoshRosen commented:
Also: if my above reasoning is right and this optimization is incorrect, then it's concerning that it didn't cause a test failure.  My hunch is that we don't have unit tests for the particular combinations of RDD dependency graphs, caching states, and map output availability that would expose this issue. It would be nice to write a failing regression test which would have caught the problems in the current version of this patch, since that will help us to gain confidence that the new optimizations are safe.

ueshin committed:
Revert "Suppress extra calling getCacheLocs."

This reverts commit 9a80fad6b2f5d73e206aacf72cd5e075fa8bcba0.

ueshin committed:
Get cached locations from block manager only if the storage level of the RDD is not StorageLevel.NONE.

ueshin committed:
Put the condition that checks if the RDD has uncached partition or not into variable for readability.

ueshin commented:
@JoshRosen Thank you for your details.
It is exactly that I was noticed yesterday.
I'm modifying `DAGScheduler` and adding the tests.
I'll push the next version as soon as possible.

ueshin committed:
Add an unit test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33402/consoleFull) for   PR 6352 at commit [`b9c835c`](https://github.com/apache/spark/commit/b9c835c5c627faee42837b73845588b0bf6a9039).

SparkQA commented:
[Test build #33402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33402/consoleFull) for   PR 6352 at commit [`b9c835c`](https://github.com/apache/spark/commit/b9c835c5c627faee42837b73845588b0bf6a9039).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33402/
Test FAILed.

ueshin committed:
Fix scalastyle.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33403/consoleFull) for   PR 6352 at commit [`6f3125c`](https://github.com/apache/spark/commit/6f3125ce48413e954f92f0cd126a27cdae6ccb80).

SparkQA commented:
[Test build #33403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33403/consoleFull) for   PR 6352 at commit [`6f3125c`](https://github.com/apache/spark/commit/6f3125ce48413e954f92f0cd126a27cdae6ccb80).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33403/
Test PASSed.

ueshin commented:
I pushed and the test passed.
@JoshRosen, @srowen, Could you please take a look at this PR again?
Thanks.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -191,6 +191,11 @@ class DAGScheduler(
   def getCacheLocs(rdd: RDD[_]): Seq[Seq[TaskLocation]] = cacheLocs.synchronized {
+    // Note: if the storage level is NONE, we don't need to get locations from block manager.
+    if (rdd.getStorageLevel == StorageLevel.NONE) {
+      return Seq.fill(rdd.partitions.size)(Nil)
```
As a general style note, I'd try to avoid using `return` in Scala code, since there are some corner-cases where using it can lead to exception-handling issues (plus it results in slightly inefficient code which uses exceptions for control flow).

JoshRosen commented:
Thanks for adding that test.  This patch looks like it's in pretty good shape, but before we consider merging there's one or two other minor corner-cases that I'd like to explore.

In the current implementation of `getCacheLocs`, we first check to see whether the RDD's cache locations have been previously fetched; if so, we return the "cached" set of cache locations, and otherwise we fetch the set of locations from the block manager and store it in the cache.  This patch's optimization takes place _prior_ to checking / updating the `cacheLocs` map, meaning that it might slightly change behavior.  Specifically, I'm wondering what happened in the old code if we called `getCahceLocs()` on an RDD that wasn't cached, then cached the RDD , forced it to be computed, then called `getCacheLocs()` again as part of a different job.  In the old code, an empty set of cache locations would have been stored in the map on the first call, so I don't think the second call would see an updated set of cache locations unless we cleared the cache locations.  In the new code, non-cached RDDs won't ever cause entries to be stored in `cacheLocs`, so it's possible that the effects of caching might become visible sooner after this patch than they would in the old code.  This might be safe, but if we make this change then it should be deliberate / knowingly.  If we want to be really conservative, maybe we should move the `storageLevel` check inside the `if (!cacheLocs.contains(rdd.id)` block in order to better preserve the old behavior.

JoshRosen commented:
Actually, it looks like we end up calling `clearCacheLocs()` when submitting a new job, so the change described above probably doesn't make a difference.  To be safe, though, and to eliminate the `return`, let's go ahead and move it into the `cacheLocs`-updating block.  Once we do that, I think this will be good to go, but I'll probably pull in a scheduler maintainer for a final spot-check / review.

JoshRosen commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -318,7 +318,7 @@ class DAGSchedulerSuite
 
   test("cache location preferences w/ dependency") {
-    val baseRdd = new MyRDD(sc, 1, Nil)
+    val baseRdd = new MyRDD(sc, 1, Nil).cache()
```
To clarify for other reviewers, I think that we need these `cache()` calls so these other tests don't fail due to the skipping of the cached locations lookups.

ueshin committed:
Move the storageLevel check inside the if (!cacheLocs.contains(rdd.id)) block.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33422 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33422/consoleFull) for   PR 6352 at commit [`d858b59`](https://github.com/apache/spark/commit/d858b59928cbc173971406737a672c33524487f2).

ueshin commented:
@JoshRosen, Thank you for your comment.
I agree with you and moved the `storageLevel` check into the `if` block.

SparkQA commented:
[Test build #33422 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33422/consoleFull) for   PR 6352 at commit [`d858b59`](https://github.com/apache/spark/commit/d858b59928cbc173971406737a672c33524487f2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33422/
Test PASSed.

JoshRosen commented:
LGTM.  /ping @markhamstra or @kayousterhout for final sign-off on scheduler-related changes.

markhamstra commented:
LGTM

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -342,6 +342,35 @@ class DAGSchedulerSuite
+      Seq(makeBlockManagerId("hostA"), makeBlockManagerId("hostB"))
+    val jobId = submit(rddE, Array(0))
+    val finalStage = scheduler.jobIdToActiveJob(jobId).finalStage
+    assert(scheduler.getMissingParentStages(finalStage).size === 0)
```
Why isn't D a missing parent stage here?

kayousterhout replied:
It looks like what happens is that the call to submit() causes the first set of missing parent stages to be submitted, so at that point, stage D is submitted.  Can you add a comment describing this?

JoshRosen replied:
Since there's a one-to-one dependency from D to E, won't D and E be computed in the same stage?

kayousterhout replied:
Ah I see.  What if we changed this test to, instead of directly calling getMissingParentStages, just directly inspect DAGScheduler.runningStages (since that's already private[scheduler]) to make sure it contains the one stage we expect? I'd find that more intuitive, since that more directly tests the underlying issue we're trying to verify.

JoshRosen replied:
That's a good idea; let's do this.

ueshin replied:
@kayousterhout, Thank you for your checking this PR.
I see, and should I revert `getMissingParentStages` to `private` ?

markhamstra replied:
Yes, if we're not going to use it in the test suite, then it should go back to private.

ueshin replied:
Ah, I found that only checking if the `DAGScheduler.runningStages` contains one stage is not enough because it also contains one stage including `A` if the `C` is not cached yet.
I think we should also check the size of the final stage's missing parents.

kayousterhout replied:
I was thinking you could inspect the contents of the stages in runningStages to make sure the Id is correct

Sent from my iPhone

> On May 26, 2015, at 7:53 PM, Takuya UESHIN <notifications@github.com> wrote:
> 
> In core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
> 
> > +   * If the RDD C, which has only one ancestor via a narrow dependency, is cached, then we won't
> > +   * need to compute A, even if it has some unavailable output partitions. The same goes for B:
> > +   * if B is 100% cached, then we can avoid the shuffle on A.
> > +   */
> > +  test("SPARK-7826: regression test for getMissingParentStages") {
> > +    val rddA = new MyRDD(sc, 1, Nil)
> > +    val rddB = new MyRDD(sc, 1, List(new ShuffleDependency(rddA, null)))
> > +    val rddC = new MyRDD(sc, 1, List(new OneToOneDependency(rddB))).cache()
> > +    val rddD = new MyRDD(sc, 1, Nil)
> > +    val rddE = new MyRDD(sc, 1,
> > +      List(new OneToOneDependency(rddC), new OneToOneDependency(rddD)))
> > +    cacheLocations(rddC.id -> 0) =
> > +      Seq(makeBlockManagerId("hostA"), makeBlockManagerId("hostB"))
> > +    val jobId = submit(rddE, Array(0))
> > +    val finalStage = scheduler.jobIdToActiveJob(jobId).finalStage
> > +    assert(scheduler.getMissingParentStages(finalStage).size === 0)
> Ah, I found that only checking if the DAGScheduler.runningStages contains one stage is not enough because it also contains one stage including A if the C is not cached yet.
> I think we should also check the size of the final stage's missing parents.
> 
> —
> Reply to this email directly or view it on GitHub.
>

ueshin replied:
Ah, the `runningStages` contains one stage and it's id is `1`, right?

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -342,6 +342,35 @@ class DAGSchedulerSuite
+   * need to compute A, even if it has some unavailable output partitions. The same goes for B:
+   * if B is 100% cached, then we can avoid the shuffle on A.
+   */
+  test("SPARK-7826: regression test for getMissingParentStages") {
```
This test name is misleading, because the test isn't testing for the bug described by SPARK-7826.  Can you change this to something more descriptive like "getMissingParentStages doesn't return cached stages"?

kayousterhout replied:
Actually given my below comment, I'm not sure what this is supposed to be testing?

JoshRosen replied:
Good call on the test naming change.  This test was derived from [a comment](https://github.com/apache/spark/pull/6352#issuecomment-104847514) that I left upthread which described a case in which the logic in an earlier version of this patch was incorrect.

An older version of this patch would end up skipping the check to see whether `C` was cached, which could cause the scheduler to mistakenly think that the shuffle from A to B would need to be performed.  The earlier version of the patch passed all tests despite this logic error, so we added this test to try to exercise that case.

kayousterhout replied:
So was the previous issue triggered by the fact that E has two narrow dependencies?  Just trying to figure out the appropriate naming here for what's being tested.

JoshRosen replied:
I was considering something like "getMissingParentStages should consider all ancestor RDDs' cache statuses", since not doing that was the cause of the bug in an earlier version of this patch.  That's still a bit cryptic, though.

I guess that we don't technically need `D` to trigger the bug, so I suppose we could remove it as a simplification.

ueshin replied:
I see.
I'll change the test name and remove the `D`.

ueshin committed:
Simplify the unit test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33559/consoleFull) for   PR 6352 at commit [`10b1b22`](https://github.com/apache/spark/commit/10b1b22d710be6deed8cf0931bd22ba26b688118).

ueshin commented:
I modified the unit test.
Thank you all for your instructions.

SparkQA commented:
[Test build #33559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33559/consoleFull) for   PR 6352 at commit [`10b1b22`](https://github.com/apache/spark/commit/10b1b22d710be6deed8cf0931bd22ba26b688118).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33559/
Test FAILed.

ueshin commented:
Retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33574/
Test FAILed.

kayousterhout commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33575/consoleFull) for   PR 6352 at commit [`10b1b22`](https://github.com/apache/spark/commit/10b1b22d710be6deed8cf0931bd22ba26b688118).

SparkQA commented:
[Test build #33575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33575/consoleFull) for   PR 6352 at commit [`10b1b22`](https://github.com/apache/spark/commit/10b1b22d710be6deed8cf0931bd22ba26b688118).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33575/
Test PASSed.

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -342,6 +342,29 @@ class DAGSchedulerSuite
+      Seq(makeBlockManagerId("hostA"), makeBlockManagerId("hostB"))
+    submit(rddD, Array(0))
+    assert(scheduler.runningStages.size === 1)
+    assert(scheduler.runningStages.head.id === 1)
```
Can you actually change this to:
assert(scheduler.runningStages.head.isInstanceOf[ResultStage])?

And then add a comment saying something like "Make sure that the scheduler is running the final result stage. Because C is cached, the shuffle map stage to compute A does not need to be run."

kayousterhout replied:
(I think this is more intuitive; otherwise, it's hard for someone looking at this to understand why the ID should be 1. This also makes the test more agnostic to unrelated scheduler internals, like if we change the way we assign IDs to stages)

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -342,6 +342,29 @@ class DAGSchedulerSuite
+   * need to compute A, even if it has some unavailable output partitions. The same goes for B:
+   * if B is 100% cached, then we can avoid the shuffle on A.
+   */
+  test("SPARK-7826: getMissingParentStages should consider all ancestor RDDs' cache statuses") {
```
Can you remove "SPARK-7826" from the name of this test, since the test isn't checking for the bug described by SPARK-7826?  It's great to add this test in the PR -- but having the JIRA name in the test name is something we usually only do when the test is for the issue described by that JIRA.

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -342,6 +342,29 @@ class DAGSchedulerSuite
+   * order to compute D, we need to figure out whether the shuffle A -> B should be performed.
+   * If the RDD C, which has only one ancestor via a narrow dependency, is cached, then we won't
+   * need to compute A, even if it has some unavailable output partitions. The same goes for B:
+   * if B is 100% cached, then we can avoid the shuffle on A.
```
Josh's comment was an awesome description of how the dependencies should be computed, but isn't quite appropriate here as the comment for the test.  What about something like:

This test ensures that if a particular RDD is cached, RDDs earlier in the dependency chain are not computed.  It constructs the following chain of dependencies:
+---+ shuffle +---+    +---+    +---+
 | A |<--------| B |<---| C |<---| D |
+---+         +---+    +---+    +---+
Here, B is derived from A by performing a shuffle, C has a one-to-one dependency on B, and D similarly has a one-to-one dependency on C.  If none of the RDDs were cached, this set of RDDs would result in a two stage job: one ShuffleMapStage, and a ResultStage that reads the shuffled data from RDD A.  This test ensures that if C is cached, the scheduler doesn't perform a shuffle, and instead computes the result using a single ResultStage that reads C's cached data.

kayousterhout commented:
Just a few more comments on improving the documentation and understandability of the test.  @JoshRosen has recently pointed out that the schedule code is extremely difficult to understand and check for correctness, and I think having easily understandable and well-documented tests is a step towards making the scheduler code more friendly.

ueshin committed:
Modify a test and the documentation.

ueshin commented:
I modified what you mentioned.
I agree with the difficulties, so please let me know if there are other things I can do here.
Thanks.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33629 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33629/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).

SparkQA commented:
[Test build #33629 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33629/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33629/
Test FAILed.

ueshin commented:
Retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33631 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33631/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).

SparkQA commented:
[Test build #33631 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33631/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33631/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33637/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).

SparkQA commented:
[Test build #33637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33637/consoleFull) for   PR 6352 at commit [`3d4d036`](https://github.com/apache/spark/commit/3d4d036aecbc7ad3b0a75f2ad8de9ce866a3131e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33637/
Test PASSed.

kayousterhout commented:
LGTM

kayousterhout commented:
Thanks @ueshin ! I merged this since @JoshRosen and @markhamstra LGTM'ed a while ago.

ueshin commented:
@kayousterhout, Thank you for merging this!
